homeassistant/helpers/config_validation.py

```diff
@@ -114,6 +118,47 @@
 TIME_PERIOD_ERROR = "offset {} should be format 'HH:MM', 'HH:MM:SS' or 'HH:MM:SS.F'"
 
 
+class MustValidateInExecutor(HomeAssistantError):
+    """Raised when validation must happen in an executor thread."""
+
+
+class _Hass(threading.local):
+    """Container which makes a HomeAssistant instance available to validators."""
+
+    hass: HomeAssistant | None = None
+
+
+_hass = _Hass()
+"""Set when doing async friendly schema validation."""
+
+
+_validating_async: ContextVar[bool] = ContextVar("_validating_async", default=False)
+"""Set to True when doing async friendly schema validation."""
+
+
+def _async_get_hass_or_none() -> HomeAssistant | None:
+    """Return the HomeAssistant instance or None.
+
+    First tries core.async_get_hass_or_none, then _hass which is
+    set when doing async friendly schema validation.
+    """
+    return async_get_hass_or_none() or _hass.hass
+
+
+def not_async_friendly[**_P, _R](validator: Callable[_P, _R]) -> Callable[_P, _R]:
```

Comment:
It would be nice to keep the names similar. Maybe make the decorator must_validate_in_executor or validate_executor.

Reply:
Do you mean keep the name of the decorator similar to the name of the exception and the helper function below?

Reply:
Yes

Reply:
What do you think would be a good name? I thought naming the decorator something like `validate_in_executor` would be misleading since the implementation currently only enforces that if the `cv.async_validate` is used to validate.
But maybe it's OK if the decorator is documented?

Reply:
> What do you think would be a good name? I thought naming the decorator something like `validate_in_executor` would be misleading since the implementation currently only enforces that if the `cv.async_validate` is used to validate. But maybe it's OK if the decorator is documented?

Thanks for giving some more color on the naming choice.  `@blocking`, `@blocking_io`, `@not_async_safe`, `@blocks`... naming this one is hard... `@not_async_friendly` is fine.